File: swift/DesignPatterns/Structural/Proxy.swift
//
//  Proxy.swift
//  DesignPatterns
//
//  Created by Soe Min Thein on 23/05/2026.
//

import Foundation

protocol DatabaseDAO {
    func query()
}

struct DatabaseDAOImpl: DatabaseDAO {
    func query() {
        print("QUERY")
    }
}

struct CacheProxyDatabase: DatabaseDAO {
    private let dao: DatabaseDAO
    
    init(dao: DatabaseDAO) {
        self.dao = dao
    }
    
    func query() {
        print("Cache:: Query from Cache")
        dao.query()
    }
}

struct LoggingProxy: DatabaseDAO {
    private let dao: DatabaseDAO
    
    init(dao: DatabaseDAO) {
        self.dao = dao
    }
    
    func query() {
        print("Log:: Start Query")
        dao.query()
        print("Log:: End Query")
    }
}

func proxyExample() {
    let db = DatabaseDAOImpl()
    let cache = CacheProxyDatabase(dao: db)
    let logProxy = LoggingProxy(dao: cache)
    
    logProxy.query()
}

File: dart/design_patterns/lib/structural/proxy.dart
// Subject

abstract class DatabaseDAO {
  void query();
}

// Real Subject

class DatabaseDAOImpl implements DatabaseDAO {
  @override
  void query() {
    print('QUERY');
  }
}

// Cache Proxy

class CacheProxyDatabase implements DatabaseDAO {
  final DatabaseDAO _dao;

  CacheProxyDatabase({required DatabaseDAO dao}) : _dao = dao;

  @override
  void query() {
    print('Cache:: Query from Cache');
    _dao.query();
  }
}

// Logging Proxy

class LoggingProxy implements DatabaseDAO {
  final DatabaseDAO _dao;

  LoggingProxy({required DatabaseDAO dao}) : _dao = dao;

  @override
  void query() {
    print('Log:: Start Query');
    _dao.query();
    print('Log:: End Query');
  }
}

// Usage

void main() {
  final db = DatabaseDAOImpl();
  final cache = CacheProxyDatabase(dao: db);
  final logProxy = LoggingProxy(dao: cache);

  logProxy.query();
}
